hera_cal/abscal.py

```diff
@@ -1291,6 +1295,280 @@ def global_phase_slope_logcal(model, data, antpos, reds=None, solver='linfit', w
+                assert numerator < max_numerator, 'Could not find a reasonable multiplier to put this array on an integer grid.'
+            
+            # Look for divisors compress the array but keep every antenna on an integer grid
+            for denominator in range(max_numerator, 0, -1):
+                if np.max(np.abs(np.rint(this_dim_positions * numerator / denominator) - this_dim_positions * numerator / denominator)) < tol:
+                    break
+            
+            # Apply ratio found
+            for ap in transformed_antpos:
+                transformed_antpos[ap][dim] *= (numerator / denominator)
+
+def complex_phase_abscal(data, model, reds, data_bls, model_bls):
+    """
+    Calculates gains that would absolute calibrate already redundantly-calibrated data.  
+    Only operates one polarization at a time. 
+    
+    Parameters:
+    ----------
+    data : DataContainer or RedDataContainer
+        Dictionary-like container mapping baselines to data visibilities to abscal
+    model : DataContainer or RedDataContainer
+        Dictionary-like container mapping baselines to model visibilities
+    reds : list of lists
+        List of list of redundant baselines tuples like (0, 1, 'ee')
+    data_bls : list of tuples
+        List of baselines in data to use.
+    model_bls : list of tuples
+        List of baselines in model to use. Must correspond the same physical separations as data_bls.
+        
+    Returns:
+    -------
+    meta : dictionary
+        Contains keys for 'Lambda_sol' (degneracy solutions) and 'Z_sol' (solution qualities)
+    delta_gains : dictionary
+        Dictionary mapping antenna keys like (0, 'Jee') to gains of the same shape of the data
+    """
+    # Check that baselines selected are for the same polarization
+    pols = list(set([bl[2] for bls in (data_bls, model_bls) for bl in bls]))
+    assert len(pols) == 1, 'complex_phase_abscal() can only solve for one polarization at a time.'
+    
+    # Get transformed antenna positions and baselines
+    transformed_antpos = redcal.reds_to_antpos(reds)
+    _put_transformed_array_on_integer_grid(transformed_antpos)
+    transformed_b_vecs = np.rint([transformed_antpos[jj] - transformed_antpos[ii] for (ii,jj,pol) in data_bls]).astype(int)
+    
+    # Get number of baselines and times/freqs
+    Ngroups = len(data_bls)
+    Ntimes, Nfreqs = data[data_bls[0]].shape
+
+    # Get normalized data/model ratios and weights
+    normalized_data_model_ratio = np.zeros((Ntimes, Nfreqs, Ngroups), dtype=complex)
+    weights = np.ones((Ntimes, Nfreqs, Ngroups), dtype=float)
+    for nn in range(Ngroups):
+        Vhat_n = data[data_bls[nn]]
+        Vbar_n = model[model_bls[nn]]
+        r_n = Vhat_n / Vbar_n
+        normalized_data_model_ratio[:, :, nn] = r_n / np.abs(r_n)
+        weights[:, :, nn] = np.abs(Vbar_n * Vhat_n)
+
+    # Get solution for degenerate phase slopes
+    weights /= np.mean(weights, axis=-1, keepdims=True)
```

Comment:
For ease of reading and logical flow, I'd put this line with the previous block

Reply:
Yeah, that makes sense. I'll move that line up

Reply:
This bit is gone now.